File: cpp/0002.add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2)
    {
        ListNode *res = new ListNode(0);
        ListNode *p = l1;
        ListNode *q = l2;
        ListNode *r = res;
        ListNode *last = NULL;
        int add = 0;
        while (p || q || add)
        {
            int x = add;
            if (p)
            {
                x += p->val;
                p = p->next;
            }
            if (q)
            {
                x += q->val;
                q = q->next;
            }
            add = x / 10;
            x %= 10;
            r->val = x;
            r->next = new ListNode(0);
            last = r;
            r = r->next;
        }
        if (r != res && r->val == 0)
        {
            delete r;
            last->next = NULL;
        }
        return res;
    }
};

File: rust/0002_add-two-numbers.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution
{
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>>
    {
        if l1.is_none() && l2.is_none()
        {
            return None;
        }

        let mut s = 0;
        let mut a = &l1;
        let mut b = &l2;
        let mut head = None;
        let mut cur = &mut head;

        while s > 0 || !a.is_none() || !b.is_none()
        {
            if let Some(p) = a {
                s += p.val;
                a = &p.next;
            }

            if let Some(p) = b {
                s += p.val;
                b = &p.next;
            }

            cur.replace(Box::new(ListNode {
                val: s % 10,
                next: None,
            }));

            if let Some(p) = cur {
                cur = &mut p.next;
            }

            s /= 10;
        }

        return head;
    }
}
